hpat/hiframes/dataframe_pass.py

```diff
@@ -1357,6 +1372,31 @@ def _run_call_drop(self, assign, lhs, rhs):
         _init_df = _gen_init_df(out_typ.columns)
         return self._replace_func(_init_df, data, pre_nodes=nodes)
 
+    def _run_call_df_isna(self, assign, lhs, rhs):
+        df_var = rhs.args[0]
+        df_typ = self.typemap[df_var.name]
+
+        # impl: for each column, convert data to series, call S.isna(), get
+        # output data and create a new dataframe
+        n_cols = len(df_typ.columns)
+        data_args = tuple('data{}'.format(i) for i in range(n_cols))
+
+        func_text = "def _isna_impl({}):\n".format(", ".join(data_args))
+        for d in data_args:
+            func_text += "  {} = hpat.hiframes.api.init_series({})\n".format(d + '_S', d)
+            func_text += "  {} = hpat.hiframes.api.get_series_data({}.isna())\n".format(d + '_O', d + '_S')
+        func_text += "  return hpat.hiframes.pd_dataframe_ext.init_dataframe({}, None, {})\n".format(
+            ", ".join(d + '_O' for d in data_args),
+            ", ".join("'{}'".format(c) for c in df_typ.columns))
```

Comment:
I would propose to create list of lines. Then join the list using `\n`. E.g.:
`func_lines = ['first_line']`
`func_lines.append('second_line')`
`func_lines.append('third_line')`
`func_text = '\n'.join(func_lines) # func_text = os.linesep.join(func_lines)` ([os.linesep](https://docs.python.org/3/library/os.html#os.linesep))

When concatenating strings in Python new string is created each time. E.g. here 5 strings will be created (`he`, `ll`, `hell`, `o`, `hello`):
`a = 'he'`
`a += 'll'`
`a += 'o'`

Reply:
Good point, rewritten it in the last commit.